[16.0] add base_external_system_ms_client_assertion

NL66278 opened the pull request.

NL66278 committed:
[ADD] base_external_system_oauth: new module

NL66278 committed:
[ADD] base_external_system_http: new module

NL66278 committed:
[IMP] *_system_http: streamlined logging

NL66278 committed:
[IMP] base_external_system_oauth: pre-commit stuff

NL66278 committed:
[IMP] base_external_system_http: pre-commit stuff

NL66278 committed:
[MIG] base_external_system_http: Migration to 16.0

NL66278 committed:
[MIG] base_external_system_oath: migration to 16.0

NL66278 committed:
[MIG] base_external_system: re-apply 10.0 changes

NL66278 committed:
[IMP] *-system_oauth: provide tests

NL66278 committed:
[ADD] base_external_system_ms_client_assertion: new module

NL66278 committed:
[TST] ..client_assertion: remove for now actual connection test

NL66278 committed:
[FIX] base_..._oauth: no need for basic type provider.

Basic authentication is no oauth at all and fully handled by base_external_system_http.

ntsirintanis commented on base_external_system_oauth/models/auth_oauth_provider.py:
```diff
@@ -0,0 +1,88 @@
+            ("demo", "DEMO Just for testing purposes"),
+            ("microsoft_client_secret", "Microsoft Client Secret Authentication"),
+        ],
+        default="basic",
```
Bug here, it should be 'demo'

NL66278 replied:
Ah, yes, I removed the useless (in this context) basic method, but forgot to adapt this one.

NL66278 committed:
[IMP] base_..assertion: optionally use private_key value as is

In some installations, especially odoo.sh, it is not possible for the Odoo admin to put
files like the private key in the filesystem. In that case the private key can be pasted
into the private key field, and the value will be used and not interpreted as a filename.

NL66278 committed:
[IMP] base_...assertion: add the xt5 header

ntsirintanis committed:
[FIX] retrieve headers from kwargs

ntsirintanis committed:
[FIX] base_external_system_ms_client_assertion: type in _set_headers

ntsirintanis committed:
[FIX] base_external_system_oauth: add headers in kwargs

NL66278 committed:
[FIX] base_..._oauth: more pre-commit stuff

NL66278 commented:
For some time now successfully running in production at large customer.

NL66278 committed:
[FIX] ..oauth: problem with setting token variable

NL66278 committed:
[FIX] base_external_system: previous_dir now also property

NL66278 committed:
[FIX] base_external..: should be possible to override timeout

NL66278 committed:
[FIX] base_external..: prevent excessive logging

NL66278 committed:
[FIX] restore basic oauth connection

NL66278 commented:
@thomaspaulb Could you add a no stale label here?

lfreeke reviewed (approved):
:+1:

This have been working without issues for our customer

thomaspaulb commented:
@NL66278 Sorry, this can't be merged in its current form. It's too many things at once, and covers 4 different modules.

Can we start by opening up a PR to `base_external_system`, with the refactors that are being done there? Add a description to the PR with checkboxes for each of the things that are added, and the source PR where it comes from (I can't find it, at least it was never merged)

NL66278 commented:
@thomaspaulb There was this one https://github.com/OCA/server-backend/pull/281 that was also never merged. All of this is running in production, alt multiple clients, without any problems.

NL66278 commented:
@thomaspaulb OK, so here is the first PR for the base module. When that is merged, I will provide, one by one, PR's for the additional modules.

NL66278 commented:
@thomaspaulb Here is the actual link to the first PR, the one to make the base module extendable,  https://github.com/OCA/server-backend/pull/414
